plugins/modeltraining/pkg/modeltraining/system.go

```diff
@@ -0,0 +1,118 @@
+package modeltraining
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/rancher/opni/apis"
+	"github.com/rancher/opni/pkg/auth/cluster"
+	"github.com/rancher/opni/pkg/resources"
+	util "github.com/rancher/opni/pkg/util/k8sutil"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/cenkalti/backoff"
+	backoffv2 "github.com/lestrrat-go/backoff/v2"
+	"github.com/nats-io/nats.go"
+	opensearch "github.com/opensearch-project/opensearch-go"
+)
+
+func newNatsConnection() (*nats.Conn, error) {
+	natsURL := os.Getenv("NATS_SERVER_URL")
+	natsSeedPath := os.Getenv("NKEY_SEED_FILENAME")
+
+	opt, err := nats.NkeyOptionFromSeed(natsSeedPath)
+	if err != nil {
+		return nil, err
+	}
+
+	retryBackoff := backoff.NewExponentialBackOff()
+	return nats.Connect(
+		natsURL,
+		opt,
+		nats.MaxReconnects(-1),
+		nats.CustomReconnectDelay(
+			func(i int) time.Duration {
+				if i == 1 {
+					retryBackoff.Reset()
+				}
+				return retryBackoff.NextBackOff()
+			},
+		),
+	)
+}
+
+func getOpensearchCredentials(ctx context.Context) (username string, password string) {
```

Comment:
You should be using the helper function from the operator
```
username, password, err := helpers.UsernameAndPassword(r.ctx, r.client, cluster)
	if err != nil {
		return err
	}
```
In this case cluster is the `opsterv1.OpenSearchCluster` object which you will need to fetch from kubernetes.

Avoid returning hardcoded credentials as this is masking errors that should probably be exposed as missing details, and it will also fall over when the user changes the default admin password.

Reply:
Fetching the opensearch object is where you want to use your retried logic.  If you get an IsNotFound error you should retry, as this won't exist until the logging backend is created.